src/input/keyboard/mod.rs

```diff
@@ -710,7 +714,11 @@ impl<D: SeatHandler + 'static> KeyboardHandle<D> {
         F: FnOnce(&mut D, &ModifiersState, KeysymHandle<'_>) -> FilterResult<T>,
     {
         trace!("Handling keystroke");
+
+        // Ensure keymap is up to date.
         let mut guard = self.arc.internal.lock().unwrap();
+        self.change_keymap(&guard.keymap);
```

Comment:
This is incorrect place to do so. You should do that sort of thing when input goes straight into `WlKeyboard` client, thus it should be after `Filter` and `Grab` logic (if the target of the `Grab` is `WlKeyboard`).

Right now after each key with IME that does grabbing, you're sending keymap event, so the input looks more like
```
[2260896.167] wl_keyboard@25.keymap(1, fd 25, 67458)
[2260896.185] wl_keyboard@13.keymap(1, fd 19, 67458)
[2260897.689] wl_keyboard@25.keymap(1, fd 26, 67458)
[2260897.712] wl_keyboard@25.key(121, 49745, 36, 0)
[2260900.771] wl_keyboard@13.keymap(1, fd 25, 67458)
[2260900.799] wl_keyboard@13.key(120, 49745, 36, 0)
[2261204.985] wl_keyboard@25.keymap(1, fd 25, 67458)
[2261205.478] wl_keyboard@13.keymap(1, fd 19, 67458)
[2261207.746] wl_keyboard@25.keymap(1, fd 26, 67458)
[2261207.766] wl_keyboard@25.key(124, 50053, 125, 1)
[2261207.775] wl_keyboard@25.modifiers(126, 64, 0, 0, 0)
[2261209.798] wl_keyboard@13.keymap(1, fd 25, 67458)
[2261209.813] wl_keyboard@13.key(123, 50053, 125, 1)
[2261209.820] wl_keyboard@13.modifiers(125, 64, 0, 0, 0)
```

And it obviously doesn't work at all, because modifiers are all messed up. The keymap change should also lead to modifiers change event for the focused client?